src/ServiceProviders/RepositoryProviders/GitLab/GitLabProvider.php

```diff
@@ -0,0 +1,246 @@
+        }
+        $this->setToken($token);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function authenticatedUser()
+    {
+        $userData = $this->gitLabAPI('user');
+        return $userData['username'];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function createRepository($local_site_path, $target, $gitlab_org = '')
+    {
+        $createRepoUrl = "api/v4/projects";
+        $target_org = $gitlab_org;
+        if (empty($gitlab_org)) {
+            $target_org = $this->authenticatedUser();
+            $postData = ['name' => $target];
+        }
+        else {
+            // We need to look up the namespace ID.
+            $group = $this->gitLabAPI('api/v4/groups/' . urlencode($gitlab_org));
+            if (!empty($group)) {
+                $postData = ['name' => $target, 'namespace_id' => $group['id']];
+            }
+            else {
+                $postData = ['name' => $target];
+            }
+        }
+        $target_project = "$target_org/$target";
+
+        // Create a GitLab repository
+        $this->logger->notice('Creating repository {repo}', ['repo' => $target_project]);
+
+        $result = $this->gitLabAPI($createRepoUrl, $postData);
+
+        // Create a git repository. Add an origin just to have the data there
+        // when collecting the build metadata later. We use the 'pantheon'
+        // remote when pushing.
+        // TODO: Do we need to remove $local_site_path/.git? (-n in create-project should obviate this need) We preserve this here because it may be user-provided via --preserve-local-repository
+        if (!is_dir("$local_site_path/.git")) {
+            $this->execGit($local_site_path, 'init');
+        }
+        // TODO: maybe in the future we will not need to set this?
+        $this->execGit($local_site_path, "remote add origin 'git@" . $this->GITLAB_URL . ":{$target_project}.git'");
+
+        return $target_project;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function pushRepository($dir, $target_project)
+    {
+        $this->execGit($dir, 'push --progress https://oauth2:{token}@{gitlab_url}/{target}.git master', ['token' => $this->token(), 'gitlab_url' => $this->GITLAB_URL, 'target' => $target_project], ['token']);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function deleteRepository($project)
+    {
+        $deleteRepoUrl = "api/v4/projects/$project";
+        $this->gitLabAPI($deleteRepoUrl, [], 'DELETE');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function projectURL($target_project)
+    {
+        return 'https://' . $this->GITLAB_URL . '/' . $target_project;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function commentOnCommit($target_project, $commit_hash, $message)
+    {
+        // @TODO -- Figure out if this is possible in GitLab API.
+    }
+
+    protected function gitLabAPI($uri, $data = [], $method = 'GET')
```

Comment:
Is this method the same for the Git provider and the CI provider?  Maybe we need an `API` namespace where we can declare service REST APIs that are used in multiple places. This could be a good place to define common `const` values.

Some thought to design probably necessary.

Reply:
I think GitLab really pushes the thought process of how things are laid out.  There is most definitely duplication between the Repository Provider and the CI Provider.  When I had forked off of the prior version of Build Tools, I ended up just having the CI provider be a stub and used the Repository piece to do all the setup.  With the larger distinction here, this makes the duplication a bit worse.

When this comment was originally posted I wasn't sure it was worthwhile to move forward with a larger design change given the more niche nature of GitLab/GitLabCI but thinking about it more, this scenario could also (I think) exist for BitBucket/BitBucket Pipelines.  So probably something we should figure out how we handle.